File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int dfs(TreeNode* root, int maxSoFar) {
        if (!root) return 0;
        int res = 0;
        if (root->val >= maxSoFar) {
            res = 1;
            maxSoFar = root->val;
        }
        res += dfs(root->left, maxSoFar);
        res += dfs(root->right, maxSoFar);

        return res;
    }
public:
    int goodNodes(TreeNode* root) {
        return dfs(root, root->val);
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.res = float("-inf")

    def dfs(self, root):
        if not root:
            return 0
        
        leftPath = max(self.dfs(root.left), 0)
        rightPath = max(self.dfs(root.right), 0)

        self.res = max(self.res, root.val + leftPath + rightPath)

        return root.val + max(leftPath, rightPath)

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.dfs(root)
        return self.res